dvc/utils/fs.py

```diff
@@ -37,13 +43,22 @@ def get_mtime_and_size(path):
                         raise
                     continue
                 size += stat.st_size
-                entry_mtime = stat.st_mtime
-                if entry_mtime > mtime:
-                    mtime = entry_mtime
+                files_mtimes[entry] = stat.st_mtime
+
+        # Why mtime for dir is actually dict_md5 from {file_path:mtime} pairs?
+        # In case of updating .dvcignore-d file in dir, mtime for directory
+        # would be updated. We don't want to detect that, yet we have to detect
+        # operations that results in dvc tracked directory mtime update and not
+        # file mtime updates (e.g. moving tracked file), hence we need to
+        # combine update mtimes with file_paths
+        mtime_signature = dict_md5(files_mtimes)
```

Comment:
The explanation is long and vague and the key point goes only in the end. 

Key point: we track both files changes and moves. So simply `max(mtime(f) for f in non_ignored_files)` won't work.